core/src/commonJvmMain/kotlin/com/ustadmobile/core/torrent/UstadTorrentManagerImpl.kt

```diff
@@ -0,0 +1,138 @@
+package com.ustadmobile.core.torrent
+
+import com.turn.ttorrent.client.*
+import com.turn.ttorrent.client.storage.FairPieceStorageFactory
+import com.ustadmobile.core.account.Endpoint
+import com.ustadmobile.core.io.ext.UstadFileCollectionStorage.createFileCollectionStorage
+import com.ustadmobile.core.util.DiTag
+import kotlinx.coroutines.*
+import org.kodein.di.DI
+import org.kodein.di.direct
+import org.kodein.di.on
+import org.kodein.di.instance
+import org.kodein.di.instanceOrNull
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import java.io.File
+
+
+interface CommunicationManagerListener {
+
+    fun onCommunicationManagerChanged(
+            meteredNetwork: Boolean
+    )
+}
+
+class UstadTorrentManagerImpl(
+        val endpoint: Endpoint,
+        override val di: DI
+) : UstadTorrentManager, CommunicationManagerListener {
+
+    private val torrentDir: File by di.on(endpoint).instance(tag = DiTag.TAG_TORRENT_DIR)
+
+    private val containerDir by di.on(endpoint).instance<File>(tag = DiTag.TAG_DEFAULT_CONTAINER_DIR)
+
+    private var containerInfoHashMap = mutableMapOf<Long, String>()
+
+    private var activeTorrentDownloadMap = mutableMapOf<Long, ActiveTorrentInfo>()
+
+ /*   private var containerDownloadPathMap = mutableMapOf<Long, String>()
+
+    private var containerManagerMap = mutableMapOf<Long, TorrentManager>()
+
+    private var containerListenerMap = mutableMapOf<Long, DownloadListenerAdapter>()*/
+
+    // on android, di uses connectionManager and a provider to get the current communicationManager
+    // on jvm, communicationManager is a singleton
+    private val communicationManager: UstadCommunicationManager?
+        get() = di.direct.instanceOrNull()
+
+    // lock add and removeTorrent
+    private val torrentLock = Mutex()
+
+    val pieceStorage: FairPieceStorageFactory by lazy {
+        FairPieceStorageFactory.INSTANCE
+    }
+
+    private fun requireCommunicationManager(): UstadCommunicationManager{
+        return communicationManager ?: throw IllegalStateException("no communicationManager found")
+    }
+
+    override suspend fun startSeeding() {
+        torrentDir.listFiles() { file: File ->
+            file.name.endsWith(".torrent")
+        }?.forEach { torrentFile ->
+            torrentLock.withLock {
+                    addTorrent(torrentFile.nameWithoutExtension.toLong(), null)
+            }
+        }
+    }
+
+    override suspend fun addTorrent(containerUid: Long, downloadPath: String?) {
+        withContext(Dispatchers.Default){
+
+            val startTime = System.currentTimeMillis()
+
+            val torrentFile = File(torrentDir, "$containerUid.torrent")
+
+            val containerFilesDir = if(downloadPath != null) File(downloadPath)
+                                    else File(containerDir, containerUid.toString())
+            containerFilesDir.mkdirs()
+            if(!torrentFile.exists())
```

Comment:
I think this a little bit of a mine in the code. There is the documentation or naming that would indicate to anyone looking at this public function that they need to make the torrent first. I think the code that creates a new torrent file should be put here.

Because this is the "Torrent Manager", it seems like a more logical place for this code to live, as opposed to the download job.